### ISSUE TITLE
Using with another template engine

### ISSUE THREAD

**Issue by slavarazum** [maizzle/framework#28]
Hi!
Maizzle its awesome framework with full control of templates.
But can we use it with other template engines?
For example if I want to use Laravel and [Blade](https://laravel.com/docs/5.8/blade) and need to include some dynamic content with specific directives like `@foreach` or include some variable. Is it possible somehow?

**Comment by cossssmin** [comment]
Hey, thanks!

That's not possible at the moment, but it is something that I would really love to have in the framework. Maybe something like [eleventy](https://github.com/11ty/eleventy/) does, where you define the template engine to use in your config.

Especially since [laravel-blade-js](https://github.com/Tarik02/laravel-blade-js) looks pretty good!

I've never created such an interface and will need to look into it, but would love to get some help with it. 

Maybe a project to consider for v1.0.0 😊

**Comment by cossssmin** [comment]
Just to make sure, I should note that everything you're describing _is_ currently possible with [Nunjucks](https://mozilla.github.io/nunjucks/), which Maizzle uses. But yeah, would be awesome if people could use their favourite templating engine.

**Comment by StevenLangbroek** [comment]
Hey @cossssmin! I have some experience in this area, and I have to say, this is *a lot* of work. It'll require fundamental changes to the documentation, cli and framework. Not to say this isn't a desirable feature, just want to prepare you for the effort. Roughly speaking, you'd need:

1. Define an interface that any templating language can abide by, something along the lines of:
```ts
interface Renderer<T, C> {
 	(
		data: T extends object,
		context: C extends object,
		maizzleOptions: any
	): Promise<HTMLString>;
}

interface TemplateEngine {
	constructor(maizzleOptions);
	/**
	 * Promise due to for example cache warming, file system discovery etc.
	 */
	register: (maizzleOptions): Promise<Renderer>;
}
```
By using promises in register + render you open the door for potentially async renderers (like Dust.js).\
2. Reimplement Nunjucks renderer to accomodate this interface, work out the kinks
3. Write docs on how to implement your own
4. Adjust CLI with new `templates=[language]` option
5. Figure out how to document this. Default to nunjucks, with toggles per code snippet? Note on top of every page?

This has the potential to drastically impact developer experience, for good or for bad depends on how much effort you spend on making sure this is thoroughly considered instead of bolted on.

**Comment by StevenLangbroek** [comment]
Another consideration is; how will this interact with email tools that have their own template syntax (e.g., Mailchimp)

**Comment by slavarazum** [comment]
@cossssmin 
>Just to make sure, I should note that everything you're describing is currently possible with Nunjucks, which Maizzle uses. But yeah, would be awesome if people could use their favourite templating engine.

My question is not about being able to use another template engine. The question is how to be able to pass dynamic data to maizzle templates? For example: user name, top articles, cart products, etc...
I can generate Maizzle email templates with static data, but how to use it in real application? Just take production Maizzle template build and paste it into Blade templates (separated to same layout structure)? In this case it's difficult to maintain it. But I don't see another way.

**Comment by cossssmin** [comment]
@slavarazum oh, you can use Blade syntax in the template, so you then have Laravel pass your email whatever data you need.

1. In your Maizzle env config, [change the output file extension](https://maizzle.com/docs/build-paths/#extension). You might even want to change the [destination path](https://maizzle.com/docs/build-paths/#path), so it outputs templates right in your Laravel app's folder, where you need them.

2. Use Nunjucks' `{% raw %}` tag to write Laravel Blade syntax in your template. For example:

```handlebars
---
title: Using Laravel Blade syntax in Maizzle
---

{% raw %}
<table>
  <tr>
    <td>Use Laravel Blade syntax, content here isn't parsed by Nunjucks</td>
  </tr>
  @foreach ($users as $user)
    <tr>
      <td>This is user {{ $user->id }}</td>
    </tr>
  @endforeach
</table>
{% endraw %}
```

**Comment by cossssmin** [comment]
@StevenLangbroek thanks for that! Yes, it's a lot of work for sure.

Some thoughts:

- definitely async
- reimplement Nunjucks might also impact current Markdown renderer, which makes me think if this should get an interface as well...
- docs no problem
- would probably default to Nunjucks as it's pretty solid

I wouldn't add it as a flag for the build command though; instead, I would want it to be configurable from the project config, something like:

```js
build: {
  templates: {
    engine: 'njk' // or blade.js, pug, handlebars, etc
  }
}
```

But yeah, I think the first step will be a test suite for the current implementation. 

Need to write those darn tests 🙈

**Comment by cossssmin** [comment]
Closing this for now, will reconsider at a later stage.

**Comment by slavarazum** [comment]
@cossssmin How actually do you use it with Laravel?
Can't find clean solution, the main problem  - we should compile it with blade additionally and have ability to build it with default Maizzle commands.

**Comment by cossssmin** [comment]
I'd just use the build options ([docs](https://maizzle.com/docs/build-config/#destination)) to output `.blade.php` files in the mailables directory. 

Compiling with Blade is not Maizzle's concern, but you can [ignore](https://maizzle.com/docs/templates/#curly-brace-conflicts) expressions inside `{{ }}` curly braces, so that Blade can process them further.

**Comment by slavarazum** [comment]
@cossssmin As I understand you are writing templates on PostHTML + Blade.
Are you using some `.blade.html` extension for input views and does it have IDE support?
What about Blade directives, wrap it with `<raw></raw>` tag or add ignore to posthtml options in config?

**Maizzle CLI** requires to have `node_modules` in root directory of maizzle project, so we are forced to have separate **package.json** and `node_modules` directory from a main project. Did I miss something, do we have ability to place maizzle project in subfolder of Laravel project?

Sorry, so many questions 🙂 Trying to find the best way of integration with Laravel.

**Comment by cossssmin** [comment]
You can also require Maizzle and use the render() method instead of building everything with CLI commands.

https://maizzle.com/docs/nodejs/

**Comment by slavarazum** [comment]
Yep, thanks.
But how to pass php/blade variables to posthtml components? It doesn't work with `locals`.

P.S. I feel like I should write an article about integration of Maizzle to other frameworks. Maybe it would be great to have this topic in Maizzle docs. It's really clear how to use Maizzle as separate project and could be very tricky to understand how to integrate it to you current application.

**Comment by cossssmin** [comment]
Haven't tested, but maybe it needs expression ignoring, i.e. `@{{ }}`?

```html
<component src="src/components/button.html" locals='{ "href": "@{{ $model->url }}" }'>
```

Note that you need `v1.0.9` for expression ignoring to work properly.

Re article: it would be awesome, we could add it to the [guides](https://maizzle.com/guides/). If you're comfortable with it, you can clone the [docs repo](https://github.com/maizzle/docs) and open a PR - guides are [here](https://github.com/maizzle/docs/tree/master/content/guides), and I'll help out with any questions you might have. You can also simply link to a text/md file and I'll add it instead, whatever works for you 👍

**Comment by cossssmin** [comment]
OK, played around with it a bit and the above won't work. It _can_ work, but it's not very obvious (or pretty):

First, you'd need to pass in the variable without the curly braces:

```html
<component src="src/components/button.html" locals='{ "src": "$model->url" }'>
```

In `button.html`:

```html
<raw>{{ </raw>{{{ src }}}<raw> }}</raw>
```

Explanation:

1. We use the [`<raw>`](https://maizzle.com/docs/tags/#raw) tag to output the left `{{ `
2. The `src` variable is ouput with `{{{ }}}`, so we [don't escape](https://github.com/posthtml/posthtml-expressions#unescaped-locals) its contents
3. We use `<raw>` again, to output the closing ` }}`

Just tested and it works, `{{ $model->url }}` is correctly output.

**Comment by slavarazum** [comment]
I will open a PR to docs repo with detailed instructions, but at first it would be great if we will find a better of Maizzle integration. I will make a topic titles soon which should be covered on my opinion.

Thanks for the solution, I have tried it inside html tag it works fine and it doesn't work in html attributes.
I have tried to write simple button component with `a` tag and `href` attribute.

IMO, this solution looks dirty:
```html
<raw>{{ </raw>{{{ src }}}<raw> }}</raw>
```
I think we can find a better way which will be work in both situations and will look more clear.
